scheds/rust/scx_p2dq/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1006 @@
+
+	if (!(taskc = lookup_task_ctx(p)))
+		return;
+
+	taskc->weight = weight;
+}
+
+
+void BPF_STRUCT_OPS(p2dq_dispatch, s32 cpu, struct task_struct *prev)
+{
+	struct cpu_ctx *cpuc;
+	struct llc_ctx *llcx;
+	u64 dsq_id = 0;
+	int i;
+
+	if (!(cpuc = lookup_cpu_ctx(cpu)) ||
+	    !(llcx = lookup_llc_ctx(cpuc->llc_id)))
+		return;
+
+	if (prev && keep_running(cpuc, prev))
+		return;
+
+	if (nr_dsqs_per_llc > MAX_DSQS_PER_LLC) {
+		scx_bpf_error("can't happen");
+		return;
+	}
+
+	// In the dispatch path we first figure out the load factor for each dsq.
+	s32 nr_queued = 0;
+	int max_dsq = 0;
+	u64 dsq_load = 0;
+	u64 max_load = 0;
+	bpf_for(i, 0, nr_dsqs_per_llc) {
+		nr_queued = scx_bpf_dsq_nr_queued(*MEMBER_VPTR(cpuc->dsqs, [i]));
+		if (nr_queued <= 0)
+			continue;
+
+		dsq_load = i <= nr_queued ? nr_queued : nr_queued / i;
+		if (dsq_load > max_load) {
+			max_load = dsq_load;
+			max_dsq = i;
+		}
+	}
+
+	// First try the DSQ with the most load.
+	if (max_dsq >= 0 && max_load > 0) {
+		dsq_id = *MEMBER_VPTR(cpuc->dsqs, [max_dsq]);
+		if (scx_bpf_dsq_move_to_local(dsq_id))
+			return;
+	}
+
+	// Try the last DSQ, this is to keep tasks sticky to their dsq type.
+	if (cpuc->dsq_index == 0 &&
+	    scx_bpf_dsq_move_to_local(cpuc->dsqs[cpuc->dsq_index]))
+		return;
+
+	u64 max_vtime = 0;
+	bpf_for(i, 0, nr_dsqs_per_llc) {
+		if (llcx->dsq_max_vtime[i] > max_vtime) {
+			max_vtime = llcx->dsq_max_vtime[i];
+			dsq_id = cpuc->dsqs[i];
+		}
+	}
+	if (dsq_id > 0 && dsq_id < nr_dsqs_per_llc)
+		scx_bpf_dsq_move_to_local(cpuc->dsqs[dsq_id]);
+}
+
+void BPF_STRUCT_OPS(p2dq_set_cpumask, struct task_struct *p,
+		    const struct cpumask *cpumask)
+{
+	struct task_ctx *taskc;
+
+	if (!(taskc = lookup_task_ctx(p)) || !all_cpumask)
+		return;
+
+	taskc->all_cpus =
+		bpf_cpumask_subset(cast_mask(all_cpumask), cpumask);
+}
+
+s32 BPF_STRUCT_OPS_SLEEPABLE(p2dq_init_task, struct task_struct *p,
+			     struct scx_init_task_args *args)
+{
+	struct bpf_cpumask *cpumask;
+	struct task_ctx *taskc;
+	struct cpu_ctx *cpuc;
+
+	s32 task_cpu = scx_bpf_task_cpu(p);
+
+	taskc = bpf_task_storage_get(&task_ctxs, p, 0,
+				    BPF_LOCAL_STORAGE_GET_F_CREATE);
+	if (!taskc) {
+		scx_bpf_error("task_ctx allocation failure");
+		return -ENOMEM;
+	}
+
+	if (!(cpumask = bpf_cpumask_create())) {
+		scx_bpf_error("task_ctx allocation failure");
+		return -ENOMEM;
+	}
+
+	if ((cpumask = bpf_kptr_xchg(&taskc->mask, cpumask))) {
+		bpf_cpumask_release(cpumask);
+		scx_bpf_error("task_ctx allocation failure");
+		return -EINVAL;
+	}
+
+	if (!(cpuc = lookup_cpu_ctx(task_cpu)))
+		return -EINVAL;
+
+	taskc->dsq_id = SCX_DSQ_INVALID;
+	taskc->llc_id = cpuc->llc_id;
+	taskc->dsq_index = init_dsq_index;
+	taskc->last_dsq_index = init_dsq_index;
+	taskc->runnable = true;
+
+	if (!all_cpumask) {
+		scx_bpf_error("NULL all_cpumask");
+		return -EINVAL;
+	}
+
+	bpf_rcu_read_lock();
+	if (p->cpus_ptr && all_cpumask)
+		taskc->all_cpus = bpf_cpumask_subset(cast_mask(all_cpumask),
```

Comment:
Maybe we could simply check `p->nr_cpus_allowed == nr_cpus`? And in this way we can also get rid of the RCU locking.